synapse/rest/client/v2_alpha/auth.py

```diff
@@ -128,9 +138,29 @@ async def on_POST(self, request, stagetype):
                     myurl="%s/r0/auth/%s/fallback/web"
                     % (CLIENT_API_PREFIX, LoginType.TERMS),
                 )
+
         elif stagetype == LoginType.SSO:
             # The SSO fallback workflow should not post here,
             raise SynapseError(404, "Fallback SSO auth does not support POST requests.")
+
+        elif stagetype == LoginType.REGISTRATION_TOKEN:
+            token = parse_string(request, "token")
+            authdict = {"session": session, "token": token}
+
+            success = await self.auth_handler.add_oob_auth(
+                LoginType.REGISTRATION_TOKEN, authdict, request.getClientIP()
+            )
+
+            if success:
+                html = self.success_template.render()
+            else:
```

Comment:
The auth checker in `synapse/handlers/ui_auth/checkers.py` either returns `True` or raises a `LoginError`, so this `else:` is never executed.

Basically if you put in an invalid token you get shown some JSON rather than a pretty message and an option to try again. This could maybe be fixed by capturing `LoginError`s in `add_oob_auth` like the normal authentication code does just above it.

The other fallbacks don't really get this problem because terms always succeeds and the recaptcha just needs to be filled out (not tested). Although if you go to https://example.org/_matrix/client/r0/auth/m.login.recaptcha/fallback/web?session=xxxxx and disable javascript, then you get a similar thing.

Reply:
That would work fine for the fallback, but then if you get an unknown token during UIAA no error messages will be returned.

The docstring for check_auth():

https://github.com/matrix-org/synapse/blob/c2c364f27f61bece85dc7fd17cdedc4b60b9f7af/synapse/handlers/ui_auth/checkers.py#L45

Although I think `SynapseError` should be `LoginError` as that is what's excepted here:

https://github.com/matrix-org/synapse/blob/c2c364f27f61bece85dc7fd17cdedc4b60b9f7af/synapse/handlers/auth.py#L600

Reply:
> Although I think `SynapseError` should be `LoginError` as that is what's excepted here.

I think you're right. The [PR that docstring was introduced with](https://github.com/matrix-org/synapse/pull/6105/files) also extracts the each UIAA checker into a separate class, where `check_auth` routinely raises `LoginError`.

I think I understand then. So the problem is that when calling `AuthHandler.check_ui_auth`, which in turn calls `RegistrationTokenAuthChecker.check_auth` during UIAA, we expect a `LoginError` if an invalid token was passed.

Conversely, when we hit `POST /_matrix/client/r0/auth/org.matrix.msc3231.login.registration_token/fallback/web` with an invalid token, we call `AuthHandler.add_oob_auth`, which in turn calls `RegistrationTokenAuthChecker.check_auth`. The call in `add_oob_auth` is  expecting either *something* or `None` from `check_auth`. Taking a look, it doesn't look like any auth checkers ever return `None` - so perhaps this is an invalid assumption, and `add_oob_auth` should be refactored to either return a result or raise `LoginError` / `SynapseError`, rather than `True` / `False`.

Then we can just wrap calls to `add_oob_auth` in `AuthRestServlet` in a try/catch for `LoginError`s that have a `401` error code. If so, return the respective template to the client indicating that the auth credentials provided (i.e a token) was invalid. Otherwise, just let the error get raised - and thus the client will receive JSON (which is expected as the request was malformed).

As an aside; it doesn't appear that [the spec](https://spec.matrix.org/unstable/client-server-api/#fallback) specifies what `POST /_matrix/client/r0/auth/<auth type>/fallback/web` needs to return - or even that that endpoint needs to be called. It looks to me as if Synapse just chose that as a sensible path - though it probably shouldn't as the `POST` variant isn't specced.

Reply:
Should I fix that up in a separate PR then, or just bundle it in here?

I think that POST endpoint is just used by synapse internally when submitting UIA creds from the web fallback page.

Reply:
> Should I fix that up in a separate PR then, or just bundle it in here?

A separate PR sounds good as it seems like a reasonably encapsulated bit of work. Simply merge `develop` into this branch once that PR lands.

> I think that POST endpoint is just used by synapse internally when submitting UIA creds from the web fallback page.

Yeah, it should just live on the `/_synapse/client` resource rather than `/_matrix/client` is all.

Reply:
> A separate PR sounds good as it seems like a reasonably encapsulated bit of work. Simply merge `develop` into this branch once that PR lands.

How does #10561 look?

> Yeah, it should just live on the `/_synapse/client` resource rather than `/_matrix/client` is all.

Oh ok